runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -6863,57 +6863,199 @@ void check_nic_amo(size_t size, void* object, mem_region_t* remote_mr) {
+                      void** object_v, size_t* size_v,
+                      gni_fma_cmd_type_t* cmd_v, mem_region_t** remote_mr_v) {
+
+  gni_post_descriptor_t post_desc_v[MAX_CHAINED_AMO_LEN];
+  atomic_bool post_done_v[MAX_CHAINED_AMO_LEN];
+  int cdi_v[MAX_CHAINED_AMO_LEN];
+  int vi;
+
+  for (vi=0; vi<v_len; vi++) {
+    // build up the post descriptor
+    post_desc_v[vi].type            = GNI_POST_AMO;
+    post_desc_v[vi].cq_mode         = GNI_CQMODE_GLOBAL_EVENT;
+    post_desc_v[vi].dlvr_mode       = GNI_DLVMODE_PERFORMANCE;
+    post_desc_v[vi].rdma_mode       = 0;
+    post_desc_v[vi].src_cq_hndl     = 0;
+    post_desc_v[vi].remote_addr     = (uint64_t) (intptr_t) object_v[vi];
+    post_desc_v[vi].remote_mem_hndl = remote_mr_v[vi]->mdh;
+    post_desc_v[vi].length          = size_v[vi];
+    post_desc_v[vi].amo_cmd         = cmd_v[vi];
+    post_desc_v[vi].first_operand   = opnd1_v[vi];
+
+    atomic_init_bool(&post_done_v[vi], false);
+    post_desc_v[vi].post_id = (uint64_t) (intptr_t) &post_done_v[vi];
+
+    // initiate the transaction
+    cdi_v[vi] = post_fma(locale_v[vi], &post_desc_v[vi]);
+  }
+
+  // Wait for all the transactions to complete
+  for (vi=0; vi<v_len; vi++) {
+    consume_all_outstanding_cq_events(cdi_v[vi]);
+    while (!atomic_load_explicit_bool(&post_done_v[vi], memory_order_acquire)) {
+      local_yield();
+      consume_all_outstanding_cq_events(cdi_v[vi]);
+    }
+    CQ_CNT_DEC(&comm_doms[cdi_v[vi]]);
+  }
+}
+
+static
+void do_remote_amo_V(int v_len, uint64_t* opnd1_v, c_nodeid_t* locale_v,
+                     void** object_v, size_t* size_v,
+                     gni_fma_cmd_type_t* cmd_v, mem_region_t** remote_mr_v) {
+
+#if HAVE_GNI_FMA_CHAIN_TRANSACTIONS
+
+  gni_post_descriptor_t post_desc;
+  gni_ct_amo_post_descriptor_t pdc[MAX_CHAINED_AMO_LEN - 1];
+  int vi, ci;
+
+  if (v_len <= 0)
+    return;
+
+  //
+  // Build up the base post descriptor
+  //
+  post_desc.next_descr      = NULL;
+  post_desc.type            = GNI_POST_AMO;
+  post_desc.cq_mode         = GNI_CQMODE_GLOBAL_EVENT;
+  post_desc.dlvr_mode       = GNI_DLVMODE_PERFORMANCE;
+  post_desc.rdma_mode       = 0;
+  post_desc.src_cq_hndl     = 0;
+  post_desc.remote_addr     = (uint64_t) (intptr_t) object_v[0];
+  post_desc.remote_mem_hndl = remote_mr_v[0]->mdh;
+  post_desc.length          = size_v[0];
+  post_desc.amo_cmd         = cmd_v[0];
+  post_desc.first_operand   = opnd1_v[0];
+
+  //
+  // Build up the chain of descriptors
+  //
+  for (vi=1, ci=0; vi<v_len; vi++, ci++) {
+    if (ci == 0)
+      post_desc.next_descr  = &pdc[0];
+    else
+      pdc[ci-1].next_descr  = &pdc[ci];
+    pdc[ci].next_descr      = NULL;
+    pdc[ci].remote_addr     = (uint64_t) (intptr_t) object_v[vi];
+    pdc[ci].remote_mem_hndl = remote_mr_v[vi]->mdh;
+    pdc[ci].length          = size_v[vi];
+    pdc[ci].amo_cmd         = cmd_v[vi];
+    pdc[ci].first_operand   = opnd1_v[vi];
+  }
+
```

Comment:
Think you need to do `pdc[ci].next_descr = NULL;` after the loop, to mark the end of the chain.  That's not otherwise guaranteed.

Reply:
That happens a few lines above. We set `pdc[ci].next_descr      = NULL;` and on the lines before that we update the previous pdc's next_descr to point at the current index (that logic was copied from the put_V code)

Reply:
Oh, crud, never mind.  Sorry.